include/mbgl/storage/resource_options.hpp

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+namespace mbgl {
+
+/**
+ * @brief Holds values for resource options.
+ */
+class ResourceOptions {
+public:
+    /**
+     * @brief Constructs a ResourceOptions object with default values.
+     */
+    ResourceOptions();
+    ~ResourceOptions();
+
+    /**
+     * @brief Sets the Mapbox access token - see https://docs.mapbox.com/help/how-mapbox-works/access-tokens/ for details.
+     *
+     * @param token Mapbox access token.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withAccessToken(std::string token);
+
+    /**
+     * @brief Gets the previously set (or default) Mapbox access token.
+     *
+     * @return const std::string& Mapbox access token.
+     */
+    const std::string& accessToken() const;
+
+    /**
+     * @brief Sets the API base URL. Default is https://api.mapbox.com for Mapbox.
+     *
+     * @param baseURL API base URL.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withBaseURL(std::string baseURL);
+
+    /**
+     * @brief Gets the previously set (or default) API base URL.
+     *
+     * @return const std::string& API base URL.
+     */
+    const std::string& baseURL() const;
+
+    /**
+     * @brief Sets the cache path.
+     *
+     * @param path Cache path.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withCachePath(std::string path);
+
+    /**
+     * @brief Gets the previously set (or default) cache path.
+     *
+     * @return cache path
+     */
+    const std::string& cachePath() const;
+
+    /**
+     * @brief Sets the asset path, which is the root directory from where
+     * the asset:// scheme gets resolved in a style.
+     *
+     * @param path Asset path.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withAssetPath(std::string path);
+
+    /**
+     * @brief Gets the previously set (or default) asset path.
+     *
+     * @return asset path
+     */
+    const std::string& assetPath() const;
+
+    /**
+     * @brief Sets the maximum cache size.
+     *
+     * @param size Cache maximum size in bytes.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withMaximumCacheSize(uint64_t size);
+
+    /**
+     * @brief Gets the previously set (or default) maximum allowed cache size.
+     *
+     * @return maximum allowed cache database size in bytes.
+     */
+    uint64_t maximumCacheSize() const;
+
+    /**
+     * @brief Sets the platform context. A platform context is usually an object
+     * that assists the creation of a file source.
+     *
+     * @param context Platform context.
+     * @return reference to ResourceOptions for chaining options together.
+     */
+    ResourceOptions& withPlatformContext(void* context);
```

Comment:
This is not ideal, but I do understand why you added it.

Ideally the FileSource should be a singleton and should not be aware of things like access token. The access token is a property of the Map and we give the access token to every request we make. Perhaps `Resource` should take care of building the URL.

Also, the logic inside the `DefaultFileSource` should be moved to GL Native eventually. Not saying this PR should address this, but it is an improvement on this direction.

Reply:
Agreed - I believe there are plenty of optimization opportunities in the overall file source implementation. I did experiment reducing the surface of file source as part of this task, but I decided not to pursue it in this PR to reduce the scope.

And yes - I think we'll soon move out the URL query generation step (which includes access token handling) to a separate object within the Map object.